Add TCP Receiver

jsouter opened the pull request:
Adds in the TCP rx thread required for merlin-detector.

timcnicholls commented on cpp/frameReceiver/src/FrameReceiverTCPRxThread.cpp:
```diff
@@ -0,0 +1,110 @@
+    getsockopt(recv_socket, SOL_SOCKET, SO_REUSEADDR, &buffer_size, &len);
+    LOG4CXX_DEBUG_LEVEL(1, logger_, "RX thread receive buffer size for port " << rx_port << " is " << buffer_size);
+
+    // Bind the socket to the specified port
```
This line (and the error message at line 82) refer to binding the socket - should both be changed to reflect fact that socket is connected (line 79)?

timcnicholls commented on cpp/frameReceiver/src/FrameReceiverTCPRxThread.cpp:
```diff
@@ -0,0 +1,110 @@
+}
+
+void FrameReceiverTCPRxThread::cleanup_specific_service(void)
+{
```
I think the socket be removed from the reactor and closed in the cleanup method to avoid dangling TCP connections in TIMED_WAIT states? See the ZMQ  RX thread implementation for an example.

timcnicholls reviewed (changes requested):
Looks good! A few minor comments. It would be great if there was a dummy decoder plugin and integration test added for TCP receivers too 😉

jsouter commented:
Cheers Tim, I'll get to work on those changes

jsouter commented:
Having some trouble REGISTERing a second Dummy decoder, think it's an issue with the CMake in the FR test directory which has to link the APP_SOURCES cpp files to the executable, and the REGISTER macro doesn't work when it is called twice within a single translation unit(?)

jsouter commented:
Looking at this again, have made the minor changes to error messages mentioned above and added in a remove_socket call in the TCPRxThread's cleanup_specific_service. Had some issues with getting the build to work when registering a new Dummy decoder class but now I have a DummyTCPFrameDecoder somewhat resembling the Merlin frame decoder. Will set about working on the actual unit tests soon.

jsouter commented:
Okay, I removed the *Lib.cpp files from the glob in the CMakeLists.txt as discussed and removed the namespacing of the registrations. I will note that because of recent changes to master using Lambda functions, we may need to add in a set(CMAKE_CXX_STANDARD 11) call somewhere in the CMake, which is absent from this PR.

GDYendell commented:
Thanks @jsouter!
